Chapter: Implementation (#4)
ysndr/MSc-Thesis: chapter/methodology.md

As part of the post-processing step discussed in [@sec:post-processing], the `LinearizationItem`s in the `Completed` linearization are reorderd by their occurence of the corresponding AST node in the source file.
To find items in this list three preconditions have to hold:

1. Each element has a corresponding span in the source
2. Items of different files appear ordered by `FileId`
3. Spans may overlap never intersect.
   $$\text{Item}^2_\text{start} \geq \text{Item}^1_\text{start} \land \text{Item}^2_\text{end} \leq \text{Item}^1_\text{end}$$
4. Items referring to the spans starting at the same position have to occur in the same order before and after the post-processing. 
   Concretely, this ensures that the tree-induced hierarchy is maintained, more precise elements follow broader ones

This first two properties are an implication of the preceding processes.
All elements are derived from AST nodes, which are parsed from files retaining their position.
Nodes that are generated by the runtime before being passed to the language server are either ignored or annotated with synthetic positions that are known to be in the bounds of the file and meet the second requirement.

> For all other nodes the second requirement is automatically fulfilled by the grammar of the Nickel language.

Reviewer: I'm sorry, I think I lost the original conversations and it seems you have re-ordered items since then, because now 2. is

> 2. Items of different files appear ordered by `FileId`

What was the point that is automatically ensured by the Nickel grammar? Also, your example seems to be about the dynamic of the language, which I suspect is a different matter. Grammar-wise, the JavaScript example and the corresponding Nickel translation would parse in the same way.

> For all other nodes the second requirement is automatically fulfilled by the grammar of the Nickel language.

Author: I think the comment was on whether the Nickel grammar implies that scopes are only nested but not overlapping

> For all other nodes the second requirement is automatically fulfilled by the grammar of the Nickel language.

Reviewer: Then, scopes are not grammar related, they appear later. Put differently, the parser doesn't care about scopes, all it do is to have occurrences of `x` in the example above, which could very well be translated directly to Nickel. The fact that different occurrences of `x` refer or not to the same variable semantically is part of the language operation semantics. I don't think saying "Nickel's grammar" or LALR grammars have this property or not really makes sense, to begin with.

  let -> num
  let -> var
}
```

<!-- 
\Begin{minipage}{.5\textwidth}
\centering

```{.graphviz #fig:nickel-ast-no-meta}
strict digraph { 
  let[label = "Let('x')"]
  num [label = "Num(5)"]
  var [label = "Var('x')"]

  let -> num
  let -> var
}
```
\captionof{figure}{AST of untyped code in Listing: \ref{lst:nickel-meta-untyped}}
\label{fig:nickel-meta-untyped}

\End{minipage}
\Begin{minipage}{0.5\textwidth}

```{.graphviz}
strict digraph { 
  meta [label="MetaValue", color="green", shape="box"]
  let[label = "Let('x')"]
  num [label = "Num(5)"]
  var [label = "Var('x')"]

  meta -> let
  let -> num
  let -> var
}
```

\captionof{figure}{AST of untyped code in Listing: \ref{lst:nickel-meta-typed}}
\label{fig:nickel-meta-typed}

\End{minipage}

 -->

### Nested Record Access

Nickel supports the referencing of variables which are represented as `Var` nodes that are resolved during runtime.
With records bound to a variable, a method to access elements inside that record is required.
The access of record members is represented using a special set of AST nodes depending on whether the member name requires an evaluation in which case resolution is deferred to the evaluation pass.

> While the latter prevents static analysis of any deeper element by the LSP, `StaticAccess` can be used to resolve any intermediate reference.

Author: In #2

> Notably, Nickel represents static access chains in inverse order as unary operations which in turn puts the terminal `Var` node as a leaf in the tree.

Author: should be in #2 as well

Being written in the same language (Rust[@rust]) as the Nickel interpreter allows NLS to integrate existing components for language analysis.
Complementary, NLS is tightly coupled to Nickel's syntax definition.
[Section @sec:linearization] will introduce the main data structure underlying all higher level LSP interactions and how the AST described in [@sec:nickel-ast] is transformed into this form.
Finally, the implementation of current LSP features is discussed in [@sec:lsp-server].

## Illustrative example

The example [@lst:nickel-complete-example] shows an illustrative high level configuration of a server.
Throughout this chapter, different sections about the NSL implementation will refer back to this example.

```{.nickel #lst:nickel-complete-example caption="Nickel example with most features shown"}
let Port | doc "A contract for a port number" =
  contracts.from_predicate (fun value =>
    builtins.is_num value &&
    value % 1 == 0 &&
    value >= 0 &&
    value <= 65535) in

let Container = {
  image | Str,
  ports | List #Port,
} in

let NobernetesConfig = {
  apiVersion | Str,
  metadata.name | Str,

>   replicas | #nums.PosNat

Reviewer: ```suggestion
  replicas | nums.PosNat
```

           | doc "The number of replicas"
           | default = 1,

>   containers | { _ : #Container },

Reviewer: ```suggestion
  containers | { _ : Container },
```

} in

let name_ = "myApp" in

let metadata_ = {
    name = name_,
} in

let webContainer = fun image => {
  image = image,
  ports = [ 80, 443 ],
} in

let image = "k8s.gcr.io/#{name_}" in

{
  apiVersion = "1.1.0",
  metadata = metadata_,
  replicas = 3,
  containers = {
    "main container" = webContainer image
  }

> } | #NobernetesConfig

Reviewer: ```suggestion
} | NobernetesConfig
```

[Section @sec:resolving-elements] explains how a completed linearization is accessed.

Advanced LSP implementations sometimes employ so-called incremental parsing, which allows updating only the relevant parts of an AST (and, in case of NLS, the derived linearization) upon small changes in the source.
However, an incremental LSP is not trivial to implement.
For once, NLS would not be able to leverage existing components from the existing Nickel implementation (most notably, the parser).
Parts of the nickel runtime, such as the typechecker, would need to be adapted or even reimplemented to work in an incremental way too.
Considering the scope of this thesis, the presented approach performs a complete analysis on every update to the source file.
The typical size of Nickel projects is assumed to remain small for quite some time, giving reasonable performance in practice.
Incremental parsing, type-checking and analysis can still be implemented as a second step in the future.

<!-- TODO mention lsif here? -->

### States

At its core the linearization in either state is represented by an array of `LinearizationItem`s which are derived from AST nodes during the linearization process.

> However, the exact structure of that array is differs as an effect of the post-processing.

Reviewer: ```suggestion
However, the exact structure of that array during the different phases differs as an effect of the post-processing.
```

`LinearizationItem`s maintain the position of their AST counterpart, as well as its type.
Unlike in the AST, *metadata* is directly associated with the element.
Further deviating from the AST representation, the *type* of the node and its *kind* are tracked separately.
The latter is used to represent a usage graph on top of the linear structure. 
It distinguishes between declarations (`let` bindings, function parameters, records) and variable usages.
Any other kind of structure, for instance, primitive values (Strings, numbers, boolean, enumerations), is recorded as `Structure`.

> To separate the phases of the elaboration of the linearization in a type-safe, the implementation is based on type-states[@typestate].

Reviewer: ```suggestion
To separate the phases of the elaboration of the linearization in a type-safe way, the implementation is based on type-states[@typestate].
```

Type-states were chosen over an enumeration bases approach for the additional flexibility they provide to build a generic interface.

> Thanks to the generic interface, the adaptions to Nickel to integrate NLS are expected to have almost no influence on the runtime performance of the language in an optimized build.

Reviewer: TIL the word adaption, thanks.

Author: varsågod

NLS implements separate type-states for the two phases of the linearization: `Building` and `Completed`.

building phase:
  ~ A linearization in the `Building` state is a linearization under construction.
    It is a list of `LinearizationItem`s of unresolved type, appended as they are created during a depth-first traversal of the AST.
  ~ During this phase, the `id` affected to a new item is always equal to its index in the array.
  ~ The Building state also records the definitions in scope of each item in a separate mapping.

post-processing phase:
  ~ Once fully built, a Building instance is post-processed to get a `Completed` linearization.
  ~ Although fundamentally still represented by an array, a completed linearization is optimized for search by positions (in the source file) thanks to sorting and the use of an auxiliary map from `id`s to the new index of items.

>   ~ Additionally, missing edges in the usage graph have been created and he types of items are fully resolved in a completed linearization.

Reviewer: ```suggestion
  ~ Additionally, missing edges in the usage graph have been created and the types of items are fully resolved in a completed linearization.
```

Note that only the former is defined as part of the Nickel libraries, the latter are specific implementations for NLS.

```{.rust #lst:nickel-definition-lineatization caption="Definition of Linearization structure"}
pub trait LinearizationState {}

pub struct Linearization<S: LinearizationState> {
    pub state: S,
}
```

```{.rust #lst:nls-definition-building-type caption="Type Definition of Building state"}
pub struct Building {
    pub linearization: Vec<LinearizationItem<Unresolved>>,
    pub scope: HashMap<Vec<ScopeId>, Vec<ID>>,
}
impl LinearizationState for Building {}
```

```{.rust #lst:nls-definition-completed-type caption="Type Definition of Completed state" }
pub struct Completed {
    pub linearization: Vec<LinearizationItem<Resolved>>,
    scope: HashMap<Vec<ScopeId>, Vec<ID>>,
    id_to_index: HashMap<ID, usize>,
}
impl LinearizationState for Completed {}
```

### Transfer from AST

The NLS project aims to present a transferable architecture that can be adapted for future languages.
Consequently, NLS faces the challenge of satisfying multiple goals

> 1. To keep up with the frequent changes to the Nickel language and ensure compatibility at minimal cost, NLS needs to *integrate critical functions* of Nickel's runtime

Reviewer: I don't know if this is the right place, but you can make the point that this worked well, because we butchered the syntax a lot those last 3 weeks (unified syntax, list -> array, changing comments, changing interpolation, enum types, etc.) and the LSP just kept working, basically.

Author: Maybe better to put this into #8

2. Adaptions to Nickel to accommodate the language server should be minimal not obstruct its development and *maintain performance of the runtime*.
<!-- what is more? -->

To accommodate these goals NLS comprises three different parts as shown in [@fig:nls-nickel-structure].

The `Linearizer` trait 
  ~ acts as an interface between Nickel and the language server.
    NLS implements a `Linearizer` specialized to Nickel which registers AST nodes and builds a final linearization.
Nickel's type checking implementation
  ~ was adapted to pass AST nodes to the `Linearizer`.
    Modifications to Nickel are minimal, comprising only few additional function calls and a slightly extended argument list.
A stub implementation
  ~ of the `Linearizer` trait is used during normal operation.
    Since most methods of this implementation are `no-op`s, the compiler should be able to optimize away all `Linearizer` calls in release builds.

<!-- TODO: caption -->

> ```{.graphviz #fig:nls-nickel-structure caption="Interaction of Componenets"}

Reviewer: ```suggestion
```{.graphviz #fig:nls-nickel-structure caption="Interaction of Components"}
```

digraph {
  nls [label="NLS"]
  nickel [label="Nickel"]
  als [label="Linearizer", shape=box]
  stub [label="Stub interface"]

  nls ->  nickel  [label="uses"]
  nls ->  als [label="implements"]
  stub ->  als [label="implements"]
  nickel ->  als  [label="uses"]
  nickel ->  stub  [label="uses"]
}
```

#### Usage Graph

At the core the linearization is a simple *linear* structure.
Yet, it represents relationships of nodes on a structural level as a tree-like structure.
Taking into account variable usage information adds back-edges to the original AST, yielding a graph structure.
Both kinds of edges have to be encoded with the elements in the list.

> Alas, items have to be referred to using `id`s since the index of items cannot be relied on(such as in e.g. a binary heap), because the array is reordered to optimize access by source position.

Reviewer: ```suggestion
Alas, items have to be referred to using `id`s since the index of items cannot be relied on (such as in e.g. a binary heap), because the array is reordered to optimize access by source position.
```

An AST inherently supports this logic.
A variable reference always refers to the closest parent node defining the name and scopes are naturally separated using branching.
Each branch of a node represents a sub-scope of its parent, i.e. new declarations made in one branch are not visible in the other.

When eliminating the tree structure, scopes have to be maintained in order to provide auto-completion of identifiers and list symbol names based on their scope as context.
Since the bare linear data structure cannot be used to deduce a scope, related metadata has to be tracked separately.
The language server maintains a register for identifiers defined in every scope.
This register allows NLS to resolve possible completion targets as detailed in [@sec:resolving-by-scope].

For simplicity, NLS represents scopes by a prefix list of integers.
Whenever a new lexical scope is entered, the list of the outer scope is extended by a unique identifier.

Additionally, to keep track of the variables in scope, and iteratively build a usage graph, NLS keeps track of the latest definition of each variable name and which `Declaration` node it refers to.

#### Linearizer

> The heart of the linearization the `Linearizer` trait as defined in [@lst:nls-linearizer-trait].

Reviewer: ```suggestion
The heart of the linearization is the `Linearizer` trait as defined in [@lst:nls-linearizer-trait].
```

        linearizer.retype_ident(lin, ident, src.clone());
      }
      Term::Let(ident, ..) |
      Term::LetPattern(Some(ident), ..)=> {
        let ty_let = binding_type(/* omitted */);
        linearizer.retype_ident(lin, ident, ty_let.clone());
      }
      _ => { /* omitted */ }
    }
```

While registering a node, NLS distinguishes 4 kinds of nodes.
These are *metadata*, *usage graph* related nodes, i.e. declarations and usages, *static access* of nested record fields, and *general elements* which is every node that does not fall into one of the prior categories.

##### Structures

```{.nickel #lst:nickel-simple-expr caption="Exemplary nickel expressions"}
// atoms

1
true
null

// binary operations
42 * 3
[ 1, 2, 3 ] @ [ 4, 5]

// if-then-else
if true then "TRUE :)" else "false :("

// string iterpolation
"#{ "hello" } #{ "world" }!"
```

In the most common case of general elements, the node is simply registered as a `LinearizationItem` of kind `Structure`.
This applies for all simple expressions like those exemplified in [@lst:nickel-simple-expr]

##### Declarations

In case of `let` bindings or function arguments name binding is equally simple.

> As discussed in [@sec:let-bindings-and-functions] the `let` node may contain both a name and pattern matches.

Reviewer: ```suggestion
As discussed in [@sec:let-bindings-and-functions] the `let` node may contain both a name and patterns.
```

Or  `pattern matching` ?

Author: Wouldnt it be matches at this point, once a pattern has been applied?

Reviewer: It's a nitpick, but I don't think I've seen pattern match used as a noun. Usually one speaks about pattern matching as a feature. Then something on the left hand side of a pattern match is rather a pattern. I would expect match to designate a tuple `(pattern, expr)` where the expression matches the pattern, or something like this. Like for regular expressions. But this is just my two cents.

Author: Ok lets say pattern then if match implies being associated to some expression

For either the linearizer generates `Declaration` items and updates its name register.
However, type information is available for name bindings only, meaning pattern matches remain untyped.

The same process applies for argument names in function declarations.

> Due to argument currying, NLS linearizes only a single argument/pattern at a time.

Reviewer: Suggestion: maybe add a link to a general entry about currying here, which is a well-known notion but still not for anyone outside of functional programming  (I guess the wikipedia page would do the job)

> Due to argument currying, NLS linearizes only a single argument/pattern at a time.

Author: ```suggestion
Due to argument currying[^https://en.wikipedia.org/wiki/Currying], NLS linearizes only a single argument/pattern at a time.
```

    webContainer [ label = "Var | webContainer" ]
    image [ label = "Var | image"]

    outer:f1 -> apiVersion
    outer:f2 -> metadata
    outer:f3 -> containers
    containers:f1 -> main_container
    main_container:f1 -> webContainer
    main_container:f2 -> image
}
```

[Section @sec:graph-representation] introduced the AST representation of Records.
As suggested by [@fig:nickel-record-ast], Nickel does not have AST nodes dedicated to record fields.
Instead, it associates field names with values as part of the `Record` node.
Since the language server is bound to process nodes individually, in effect, it will only see the values.
Therefore, it can not process record values at the same time as the outer record.
For the language server it is important to associate field names with their value, as it serves as name declaration.
For that reason, NLS distinguishes `Record` and `RecordField` as independent kinds of linearization items where `RecordFields` act as a bridge between the record and the value named after the field.

To maintain similarity to other binding types, NLS has to create a separate item for the field and the value.

> This also ensures, that the value can be linearized independently.

Reviewer: ```suggestion
This also ensures that the value can be linearized independently.
```

```{.graphviz #fig:nls-lin-records caption="Linearization of a record"}
digraph G {
    rankdir = LR;
    ranksep = 2;
    nodesep = .5;
    node[shape="record", fontname = "Fira Code", fontsize = 9]

    lin [ label = "<f1> | <f2> | <f3> | <f4> | <f5> ... | <f6> | <f7> | <f8> ...", width=.1]

    outer [ label = "Record" ]
    field_apiVersion [label = "RecordField |apiVersion "]
    field_containers [label="RecordField | containers"]
    field_Metadata [label = "RecordField | Metadata"]
    inner [ label = "Record" ]
    file_main_container [label="RecordField| main_containers"]

    lin:f1 -> outer
    outer -> lin:f2 [style = dashed]
    outer -> lin:f3 [style = dashed]
    outer -> lin:f4 [style = dashed]

    lin:f2 -> field_apiVersion
    field_apiVersion -> lin:f5 [style = dashed]

    lin:f6 -> inner
    inner -> lin:f7 [style = dashed]

    lin:f3 -> field_containers
    field_containers -> lin:f6 [style = dashed]

    lin:f4 -> field_Metadata
    field_Metadata-> lin:f8 [style = dashed]

    lin:f7 -> file_main_container
    file_main_container -> lin:f8 [style = dashed]
}
```

> To provide the necessary references, NLS makes used of the *scope safe* memory of its `Linearizer` implementation.

Reviewer: ```suggestion
To provide the necessary references, NLS makes use of the *scope safe* memory of its `Linearizer` implementation.
```

##### Variable Reference

The usage of a variable is always expressed as a `Var` node that holds an identifier.
Registering a name usage is a multi-step process.

First, NLS tries to find the identifier in its scope-aware name registry.
If the registry does not contain the identifier, NLS will linearize the node as `Unbound`.
In the case that the registry lookup succeeds, NLS retrieves the referenced `Declaration` or `RecordField`. The linearizer will then add a usage item in the `Resolved` state to the linearization and update the declaration's list of usages.

##### Resolution of Record Fields

The AST representation of record destructuring in [@fig:nickel-static-access] shows that accessing inner records involves chains of unary operations *ending* with a reference to a variable binding.
Each operation encodes one field of a referenced record.
However, to reference the corresponding declaration, the final usage has to be known.
Therefore, instead of linearizing the intermediate elements directly, the `Linearizer` adds them to a shared stack until the grounding variable reference is registered.

> Whenever a variable usage is linearized, NLS checks the stack for latent destructors.

Reviewer: I think destructors haven't been used until now. It could be better for the reader to either define them quickly: 

```suggestion
Whenever a variable usage is linearized, NLS checks the stack for latent destructors (record accesses).
```

Or to stick to using `access` everywhere.

Reviewer: It's your call, but I think it doesn't cost much to recall this here, as the reader may have forgotten, or dive directly into this chapter.

If destructors are present, it adds `Usage` items for each element on the stack.
Yet, because records are recursive it is possible that fields reference other fields' values.

Consider the following example [@lst:nickel-recursive-record], which is depicted in [@fig:nls-unavailable-rec-record-field]

```{.nickel #lst:nickel-recursive-record caption="Example of a recursive record"}
{
  y = {
    yy = "foo",
    yz = z,
  },
  z = y.yy
}
```

> ```{.graphviz #fig:nls-unavailable-rec-record-field caption="Example race condition in recursive records. The field `y.yz` cannot be not be referenced at this point as the `y` branch has yet to be linearized"}

Reviewer: ```suggestion
```{.graphviz #fig:nls-unavailable-rec-record-field caption="Example lock in recursive records. The field `y.yz` cannot be not be referenced at this point as the `y` branch has yet to be linearized"}
```

I would say a race condition is something a bit different.

In order to avoid interference with the main linearization, contracts are linearized using their own `Linearizer`.

### Post-Processing
Once the entire AST has been processed NLS modifies the Linearization to make it suitable as an efficient index to serve various LSP commands.

After the post-processing the resulting linearization

1. allows efficient lookup of elements from file locations
2. maintains an `id` based lookup
3. links deeply nested record destructors to the correct definitions
4. provides all available type information utilizing Nickel's typing backend

#### Sorting

Since the linearization is performed in a preorder traversal, processing already happens in the order elements are defined physically.
Yet, during the linearization the location might be unstable or unknown for different items.
Record fields for instance are processed in an arbitrary order rather than the order they are defined.
Moreover, for nested records and record short notations, symbolic `Record` items are created which cannot be mapped to a physical location and are thus placed at the range `[0..=0]` in the beginning of the file.

> Maintaining constant insertion performance and item-referencing require that the linearization is exclusively appended.

Reviewer: ```suggestion
Maintaining constant insertion performance and item-referencing requires that the linearization is exclusively appended.
```

> Each of these cases, break the physical linearity of the linearization.

Reviewer: Not totally clear what the "physical linearity" means here.

> Since the linearization is performed in a preorder traversal, processing already happens in the order elements are defined physically.
> Yet, during the linearization the location might be unstable or unknown for different items.
> Record fields for instance are processed in an arbitrary order rather than the order they are defined.
> Moreover, for nested records and record short notations, symbolic `Record` items are created which cannot be mapped to a physical location and are thus placed at the range `[0..=0]` in the beginning of the file.
> Maintaining constant insertion performance and item-referencing requires that the linearization is exclusively appended.
> Each of these cases, break the physical linearity of the linearization.

Author: ```suggestion
Since the linearization is performed in a preorder traversal, processing already happens in the order elements are defined in the file.
Yet, during the linearization the location might be unstable or unknown for different items.
Record fields for instance are processed in an arbitrary order rather than the order they are defined.
Moreover, for nested records and record short notations, symbolic `Record` items are created which cannot be mapped to the original source and are thus placed at the range `[0..=0]` in the beginning of the file.
Maintaining constant insertion performance and item-referencing requires that the linearization is exclusively appended.
Given the examples above, this breaks the linearity of the items with respect to their assigned position.
```

@yannham do you think this is clearer?

Reviewer: Yes, using "source" is clearer than "physical" indeed. I would still use something else than linearity. Presumably, you mean that it doesn't *preserve the original order* ?

Author: Ok I'll go with "original order"

NLS thus defers reordering of items.
The language server uses a stable sorting algorithm to sort items by their associated span's starting position.
This way, nesting of items with the same start location is preserved.

> Since several operations require efficient access to elements by `id`, which after the sorting does not correspond to the items index in the linearization, after sorting NLS creates an index mapping `id`s to list indices.

Reviewer: ```suggestion
Since several operations require efficient access to elements by `id`, which after the sorting does not correspond to the items index in the linearization, after sorting NLS creates an index mapping `id`s to the new actual indices.
```

As seen in [@fig:ncl-record-access], the items generated for each destructor only link to their parent item.
Yet, the root access is connected to a known declaration.
Since at this point all records are fully processed NLS is able to resolve destructors iteratively.

First NLS collects all deferred usages in a queue.
Each usage contains the *`id`* of the parent destructor as well as the *name* of the field itself represents.
NLS then tries to resolve the base record for the usage by resolving the parent.
If the value of the parent destructor is not yet known or a deferred usage, NLS will enqueue the destructor once again to be processed again later.
In practical terms that is after the other fields of a common record.
In any other case the parent consequently has to point to a record, either directly, through a record field or a variable.
NLS will then get the `id` of the `RecordField` for the destructors *name* and mark the `Usage` as `Known`
If no field with that name is present or the parent points to a `Structure` or `Unbound` usage, the destructor cannot be resolved in a meaningful way and will thus be marked `Unbound`.

#### Resolving types

> <!-- TODO: link to background section -->
> As a necessity for type checking, Nickel generates type variables for any node of the AST which it hands down to the `Linearizer`.
> <!-- TODO: example for types? -->
> In order to provide meaningful information, the Language Server needs to derive concrete types from these variables.
> The required metadata needs to be provided by the type checker.

Author: ```suggestion
<!-- TODO: link to background section -->
As a necessity for type checking, Nickel generates unification variables for any node of the AST which it hands down to the `Linearizer`.
In Nickel these variables act as type constraints which are solved later during the processing of a file.
As a result, when a node is encountered by NLS, its type is not fully resolved yet.
Similar to runtime processing, NLS needs to resolve the final types separately.
<!-- TODO: example for types? -->
```

Reviewer: Suggestion, with some rewording and also insisting on type inference. Now that I think of it, type inference is the actual reason why you don't know the types right away: in a fully explicitly typed language, you wouldn't have to solve constraints to begin with.

```suggestion
<!-- TODO: link to background section -->
Nickel features type inference in order to relieve the programmer of the burden of writing a lot of redundant type annotations. In a typed block, the typechecker is able to guess the type of all the values, even when they are not explicitly annotated by the user. To do so, the typechecker generates constraints derived from inspecting the AST, and solve them along the way. As a consequence, when a node is first encountered by NLS, its type is not necessarily  known. There, the typechecker associate to the new node a so-called unification variable, which is a placeholder for a later resolved type. This unification variable is handed down to the `Linearizer`.

Similar to runtime processing, NLS needs to resolve the final types separately. After typechecking, NLS is eventually able to query the resolved type corresponding to an AST node using the associated unification variable.
<!-- TODO: example for types? -->
```

Author: I agree its clearer wrt. to the inference

If the search has to be aborted, the query does not have a corresponding `LinearizationItem`.

```{.rust #lst:nls-resolve-at caption="Resolution of item at given position"}
impl Completed {
  pub fn item_at(
    &self,
    locator: &(FileId, ByteIndex),
  ) -> Option<&LinearizationItem<Resolved>> {
    let (file_id, start) = locator;
    let linearization = &self.linearization;
    let item = match linearization
      .binary_search_by_key(
        locator,
        |item| (item.pos.src_id, item.pos.start))
      {
        // Found item(s) starting at `locator`
        // search for most precise element
        Ok(index) => linearization[index..]
          .iter()
          .take_while(|item| (item.pos.src_id, item.pos.start) == locator)
          .last(),
        // No perfect match found
        // iterate back finding the first wrapping linearization item
        Err(index) => {
          linearization[..index].iter().rfind(|item| {
            // Return the first (innermost) matching item
            file_id == &item.pos.src_id
            && start > &item.pos.start
            && start < &item.pos.end
          })
        }
    };
    item
  }
}
```

#### Resolving by ID

> During the building process item IDs are equal to their index in the underlying List which allows for efficient access by ID.

Reviewer: ```suggestion
During the building process item IDs are equal to their index in the underlying array which allows for efficient access by ID.
```
Or `Vec`. It's a nitpick but list is a bit troubling because it reads like it is a linked list, which it surely shouldn't, given the complexity characteristics of lists.

To allow similarly efficient access to nodes with using IDs a `Completed` linearization maintains a mapping of IDs to their corresponding index in the reordered array.
A queried ID is first looked up in this mapping which yields an index from which the actual item is read.

#### Resolving by scope

During the construction from the AST, the syntactic scope of each element is eventually known.
This allows to map scopes to a list of elements defined in this scope.
Definitions from higher scopes are not repeated, instead they are calculated on request.

> As scopes are lists of scope fragments, for any given scope the set of referable nodes is determined by unifying IDs of all prefixes of the given scope, then resolving the IDs to elements.

Reviewer: It may be worth it to explain just slightly more how are scopes represented in practice. Given this sentences, it seems like they are linked lists of arrays, very similar to the `nickel-lang::environment::Environment` data structure, that are then flattened when requesting an identifier?

Author: Its a Map of arrays actually and not too dissimilar from the layers of `Envrionment`
The differnce is that layers are represented in a flat structure and persistent.
Each layer (if you will) is uniquely identified by its scope(-path), the elements of which are are the prefix I refer to.

Without thinking of a concrete Nickel example the scope could look like this:

```
1 -> [Node(1)]
1/1 -> [Node(2), Node(6)]
1/2 -> [Node(3)],
1/2/1 -> [Node(4)],
1/2/2 -> [Node(5)]
```

If we resolve the names in scope for `Node(5)` we get

1 ->  [Node(1)]
1/2 -> [Node(3)]

=> [Node(1), Node(3)]
 
In fact this should be the same as if we stored the environment for every node, just more efficient(?)

Author: @yannham 
Changed this to [15ea55c9cd2c743918e35e5734ef6cc47f52f67a](https://github.com/ysndr/MSc-Thesis/pull/4/commits/15ea55c9cd2c743918e35e5734ef6cc47f52f67a)

Do you think this makes it clearer?

Reviewer: Yes, the diagram in particular makes things clearer.
